oprf/group/suite.go

```diff
@@ -0,0 +1,201 @@
+type Suite interface {
+	// Returns the identifying name of the group
+	Name() string
+
+	// Returns the canonical (fixed) generator for defined group
+	Generator() *Point
+
+	// Returns the order of the canonical generator in the group.
+	Order() *Scalar
+
+	// Performs a transformation to encode bytes as a Element object in the
+	// group.
+	HashToGroup([]byte) (*Point, error)
+
+	// Performs a transformation to encode bytes as a Scalar object in the
+	// appropiate group.
+	// TODO: should we keep it? Seems only be used on the Verified case..
+	HashToScalar([]byte) (Scalar, error)
+
+	// Samples a random scalar value from the field of scalars defined by the
+	// group order.
+	RandomScalar() (*Scalar, error)
+
+	// Performs a scalar multiplication of the Generator with some scalar
+	// input
+	ScalarMultBase(*Scalar) (*Point, error)
+
+	// Returns the ByteLength of objects associated with the Ciphersuite
+	ByteLength() int
+}
+
+// Name returns the name associated with the chosen ciphersuite.
+func (c *Ciphersuite) Name() string {
+	return c.name
+}
+
+// Generator returns the canonical (fixed) generator for the defined group.
+func (c *Ciphersuite) Generator() *Point {
+	return &Point{c, c.curve.Params().Gx, c.curve.Params().Gy}
+}
+
+// Order returns the order of the canonical generator in the group.
+func (c *Ciphersuite) Order() *Scalar {
+	return &Scalar{c, c.curve.Params().N}
+}
+
+func getH2CSuite(c *Ciphersuite) (HashToPoint, error) {
+	var suite h2c.SuiteID
+
+	switch c.Name() {
+	case "OPRFP256-SHA512-ELL2-RO":
+		suite = h2c.P256_XMDSHA256_SSWU_RO_
+	default:
+		return nil, errors.New("invalid suite")
+	}
+
+	hasher, err := suite.Get([]byte(c.dst))
+	if err != nil {
+		return nil, err
+	}
+
+	return eccHasher{c, hasher}, nil
+}
+
+// HashToPoint produces a new point by hashing the input message.
+type HashToPoint interface {
+	Hash(msg []byte) (*Point, error)
+}
+
+type eccHasher struct {
+	suite  *Ciphersuite
+	hasher h2c.HashToPoint
+}
+
+func (h eccHasher) Hash(in []byte) (*Point, error) {
+	q := h.hasher.Hash(in)
+	x := q.X().Polynomial()
+	y := q.Y().Polynomial()
+
+	p := &Point{h.suite, new(big.Int), new(big.Int)}
+	p.x.Set(x[0])
+	p.y.Set(y[0])
+
+	// TODO: failing here
+	//if !p.IsValid() {
+	//	return nil, errors.New("invalid point")
+	//}
+
+	return p, nil
+}
+
+// HashToGroup performs a transformation to encode bytes as a Element object in the
+// group.
+func (c *Ciphersuite) HashToGroup(in []byte) (*Point, error) {
+	hasher, err := getH2CSuite(c)
+	if err != nil {
+		return nil, err
+	}
+
+	p, err := hasher.Hash(in)
+	if err != nil {
+		return nil, err
+	}
+
+	return p, nil
+}
+
+// HashToScalar performs a transformation to encode bytes as a Scalar object in the
+// appropiate group.
+func (c *Ciphersuite) HashToScalar([]byte) (*Scalar, error) {
+	return &Scalar{}, nil
+}
+
+// RandomScalar samples a random scalar value from the field of scalars defined by the
+// group order.
+// TODO: not constant time
```

Comment:
Do we intend to make it constant time soon?

Reply:
Not so soon. It is the rejection part that it is not constant.

Reply:
In a subsequent PR, we can use `hashToField` to sample scalars.